openedx/core/djangoapps/content/course_structures/tasks.py

```diff
@@ -0,0 +1,79 @@
+import json
+import logging
+
+from celery.task import task
+from opaque_keys.edx.keys import CourseKey
+from xmodule.modulestore.django import modulestore
+
+
+log = logging.getLogger(__name__)
+
+
+def generate_course_structure(course_key):
+    """
+    Generates a course structure dictionary for the specified course.
+    """
+    course = modulestore().get_course(course_key, depth=None)
+    blocks_stack = [course]
+    blocks_dict = {}
+    while blocks_stack:
+        curr_block = blocks_stack.pop()
+        children = curr_block.get_children() if curr_block.has_children else []
+        key = unicode(curr_block.scope_ids.usage_id)
+        block = {
+            "usage_key": key,
+            "block_type": curr_block.category,
+            "display_name": curr_block.display_name,
+            "children": [unicode(child.scope_ids.usage_id) for child in children]
+        }
+
+        # Retrieve these attributes separately so that we can fail gracefully if the block doesn't have the attribute.
+        attrs = (('graded', False), ('format', None))
```

Comment:
@cpennington @doctoryes @ormsbee We are seeing issues with peergrading modules. Should all modules have `format` and `graded` attributes? If so, any idea why we are seeing the error in [AN-4774](https://openedx.atlassian.net/browse/AN-4774)? If not, is the try-except block below an appropriate workaround? Should we also be concerned about the `category`, `display_name` and child-related attributes?